homeassistant/components/netgear/switch.py

```diff
@@ -0,0 +1,103 @@
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN, KEY_COORDINATOR, KEY_ROUTER
+from .router import NetgearDeviceEntity, NetgearRouter
+
+_LOGGER = logging.getLogger(__name__)
+
+
+SWITCH_TYPES = [
+    SwitchEntityDescription(
+        key="allow_or_block",
+        name="Allowed on network",
+        icon="mdi:block-helper",
+        entity_category=EntityCategory.CONFIG,
+    )
+]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up switches for Netgear component."""
+    router = hass.data[DOMAIN][entry.entry_id][KEY_ROUTER]
+    coordinator = hass.data[DOMAIN][entry.entry_id][KEY_COORDINATOR]
+    tracked = set()
+
+    @callback
+    def new_device_callback() -> None:
+        """Add new devices if needed."""
+        new_entities = []
+        if not coordinator.data:
+            return
+
+        for mac, device in router.devices.items():
+            if mac in tracked:
+                continue
+
+            new_entities.extend(
+                [
+                    NetgearAllowBlock(coordinator, router, device, entity_description)
+                    for entity_description in SWITCH_TYPES
+                ]
+            )
+            tracked.add(mac)
+
+        if new_entities:
+            async_add_entities(new_entities)
+
+    entry.async_on_unload(coordinator.async_add_listener(new_device_callback))
+
+    coordinator.data = True
+    new_device_callback()
+
+
+class NetgearAllowBlock(NetgearDeviceEntity, SwitchEntity):
+    """Allow or Block a device from the network."""
+
+    _attr_entity_registry_enabled_default = False
+
+    def __init__(
+        self,
+        coordinator: DataUpdateCoordinator,
+        router: NetgearRouter,
+        device: dict,
+        entity_description: SwitchEntityDescription,
+    ) -> None:
+        """Initialize a Netgear device."""
+        super().__init__(coordinator, router, device)
+        self.entity_description = entity_description
+        self._name = f"{self.get_device_name()} {self.entity_description.name}"
+        self._unique_id = f"{self._mac}-{self.entity_description.key}"
+        self._state = True
+        self.async_update_device()
+
+    @property
+    def is_on(self):
+        """Return true if switch is on."""
+        return self._state
+
+    async def async_turn_on(self, **kwargs):
+        """Turn the switch on."""
+        await self._router.async_allow_block_device(self._mac, ALLOW)
+
+    async def async_turn_off(self, **kwargs):
+        """Turn the switch off."""
+        await self._router.async_allow_block_device(self._mac, BLOCK)
+
+    @callback
+    def async_update_device(self) -> None:
+        """Update the Netgear device."""
+        self._device = self._router.devices[self._mac]
+        self._active = self._device["active"]
+        self._state = self._device[self.entity_description.key] == "Allow"
```

Comment:
Where is the allow state updated?

Reply:
Its part of the device_info that is collected in the async_update_device_trackers call which is performed by the coordinator.

Reply:
Do all routers provide this state?

Reply:
I don't really know but considering this part of the code:
https://github.com/MatMaul/pynetgear/blob/a7c172659945e1873c35c0f8346d1cacf84d3212/pynetgear/__init__.py#L323-L332
I guess there will be some routers that will not provide this state, for those routers the value will be set to None.
Maybe I schould change that to a default of Allow in the upstream lib?

Reply:
or maybe even better, check in this part of the code for a None vallue and set the state to True if it is None (unknown).

Reply:
Can we use the switch and block devices even if the router doesn't provide this state?

Reply:
I think the allow/block service will work even if the router doesn't provide this state.
The only problem would be how the switch is displayed inside HomeAssistant, If I hardcode the state to True (in case the router does not provide the state) the switch will always be "ON" and you can basically only block devices but not unblock them again. (although the switch.turn_on service schould still work).

What would be the best approach for this?

Reply:
Either we don't create entities for devices that don't have the state, or we set state to None (unknown) for the switch if the device doesn't report state.

Reply:
I just tested and if the state is None, both buttons are displayed which is perfect:
![afbeelding](https://user-images.githubusercontent.com/14811189/152767495-176a6e81-0011-432b-81e6-12a68f7b7866.png)

Reply:
Ok.

Reply:
Let's modify the condition here so we retain a `None` state as long as the device state is `None`.